update self hosting scaling notes (#577)
temporalio/documentation: docs/server/production-deploy.md

With that said, here are some guidelines to some common bottlenecks:

- **Database**. The vast majority of the time the database will be the bottleneck. **We highly recommend setting alerts on `ScheduleToStart` latency** to look out for this. Also check if your database connection getting saturated.
- **Internal services**. The next layer will be scaling the 4 internal services of Temporal ([Frontend, Matching, History, and Worker](https://docs.temporal.io/docs/server-architecture/)).

>   Monitor each accordingly. The Frontend service is more CPU bound, whereas the History and Matching services require more memory.

Reviewer: Any way to add more info on how to monitor each accordingly?

Author: i'd love to but dont have specific ideas apart from literally watching CPU graphs

>   If you need more instances of each service, spin them up separately with different command line arguments. You can learn more cross referencing [our Helm chart](https://github.com/temporalio/helm-charts) with our [Server Configuration reference](https://docs.temporal.io/docs/server/configuration/).

Reviewer 2: As per @mfateev comment on this [Slack Thread](https://temporalio.slack.com/archives/CTRCR8RBP/p1628981004218200): 

> They are all the same binary. So to spin up them separately you pass different command line arguments.

It would be good to mention that Frontend, Matching, History and Worker are bundled in one single binary, since the part 

>[...]the 4 *internal* services of Temporal [...]

is only implying it.

- See the **Server Limits** section below for other limits you will want to keep in mind when doing system design, including event history length.
- [Multi-Cluster Replication](https://docs.temporal.io/docs/server/multi-cluster/) is an experimental feature you can explore for heavy reads.

Finally you want to set up alerting and monitoring on Worker metrics.
When workers are able to keep up, `ScheduleToStart` latency is close to zero.
The default is 4 Workers, which should handle no more than 300 messages per second.
Scaling will depend on your workload — for example, for a Task Queue with 500 messages per second, you might want to scale up to 10 workers.
Provided you tune the concurrency of your worker / number of pollers based on your application, it should be possible to scale workers based on standard resource utilization metrics (CPU, Memory, etc).

It's possible to have too many workers.

> Monitor the poll success and poll timeout metrics:

Author: ```suggestion
Monitor the poll success (`poll_success`/`poll_success_sync`) and poll timeout metrics:
```

> Finally you want to set up alerting and monitoring on Worker metrics.
> When workers are able to keep up, `ScheduleToStart` latency is close to zero.
> The default is 4 Workers, which should handle no more than 300 messages per second.
> Scaling will depend on your workload — for example, for a Task Queue with 500 messages per second, you might want to scale up to 10 workers.
> Provided you tune the concurrency of your worker / number of pollers based on your application, it should be possible to scale workers based on standard resource utilization metrics (CPU, Memory, etc).
> 
> It's possible to have too many workers.
> Monitor the poll success (`poll_success`/`poll_success_sync`) and `poll_timeouts` metrics:
> 
> - if you see low `ScheduleToStart` latency / low percentage of poll success / high percentage of timeouts, you might have too many workers/pollers.
> - with 100% poll success and increasing `ScheduleToStart` latency, you need to scale up.

Author: ```suggestion
Finally you want to set up alerting and monitoring on Worker metrics.
When Workers are able to keep up, `ScheduleToStart` latency is close to zero.
The default is 4 Workers (aka pollers, as the Workers poll Task Queues), which should handle no more than 300 messages per second.

Specifically, the primary scaling metrics are located in the server's dynamic configs:

- `MaxConcurrentActivityExecutionSize`: Defaults to 1000
- `MaxConcurrentWorkflowTaskExecutionSize`: Defaults to 1000
- `MaxConcurrentActivityTaskPollers`: Defaults to 4
- `MaxConcurrentWorkflowTaskPollers`: Defaults to 4

Scaling will depend on your workload — for example, for a Task Queue with 500 messages per second, you might want to scale up to 10 pollers.
Provided you tune the concurrency of your pollers based on your application, it should be possible to scale them based on standard resource utilization metrics (CPU, Memory, etc).

**It's possible to have too many workers.**
Monitor the poll success (`poll_success`/`poll_success_sync`) and `poll_timeouts` metrics:

- if you see low `ScheduleToStart` latency / low percentage of poll success / high percentage of timeouts, you might have too many workers/pollers.
- with 100% poll success and increasing `ScheduleToStart` latency, you need to scale up.
```

  If you need more instances of each service, spin them up separately with different command line arguments. You can learn more cross referencing [our Helm chart](https://github.com/temporalio/helm-charts) with our [Server Configuration reference](https://docs.temporal.io/docs/server/configuration/).
- See the **Server Limits** section below for other limits you will want to keep in mind when doing system design, including event history length.
- [Multi-Cluster Replication](https://docs.temporal.io/docs/server/multi-cluster/) is an experimental feature you can explore for heavy reads.

Finally you want to set up alerting and monitoring on Worker metrics.
When Workers are able to keep up, `ScheduleToStart` latency is close to zero.
The default is 4 Workers (aka pollers, as the Workers poll Task Queues), which should handle no more than 300 messages per second.

Specifically, the primary scaling metrics are located in the server's dynamic configs:

- `MaxConcurrentActivityTaskPollers` and `MaxConcurrentWorkflowTaskPollers`: [Defaults to 4](https://github.com/temporalio/temporal/blob/fe05751305b1cb50b68efa23f8aa5f1b34f45bc5/service/worker/service.go#L121)

> - `MaxConcurrentActivityExecutionSize` and ``MaxConcurrentWorkflowTaskExecutionSize`: [Defaults to 200](https://github.com/temporalio/sdk-java/blob/bef967639fcdbe14ca37b80ac816596412846e5f/temporal-sdk/src/main/java/io/temporal/worker/WorkerOptions.java#L51)

Author: ```suggestion
- `MaxConcurrentActivityExecutionSize` and `MaxConcurrentWorkflowTaskExecutionSize`: [Defaults to 200](https://github.com/temporalio/sdk-java/blob/bef967639fcdbe14ca37b80ac816596412846e5f/temporal-sdk/src/main/java/io/temporal/worker/WorkerOptions.java#L51)
```

  If you need more instances of each service, spin them up separately with different command line arguments. You can learn more cross referencing [our Helm chart](https://github.com/temporalio/helm-charts) with our [Server Configuration reference](https://docs.temporal.io/docs/server/configuration/).
- See the **Server Limits** section below for other limits you will want to keep in mind when doing system design, including event history length.
- [Multi-Cluster Replication](https://docs.temporal.io/docs/server/multi-cluster/) is an experimental feature you can explore for heavy reads.

Finally you want to set up alerting and monitoring on Worker metrics.
When Workers are able to keep up, `ScheduleToStart` latency is close to zero.
The default is 4 Workers (aka pollers, as the Workers poll Task Queues), which should handle no more than 300 messages per second.

Specifically, the primary scaling metrics are located in the server's dynamic configs:

- `MaxConcurrentActivityExecutionSize`: Defaults to 1000
- `MaxConcurrentWorkflowTaskExecutionSize`: Defaults to 1000
- `MaxConcurrentActivityTaskPollers`: Defaults to 4

> - `MaxConcurrentWorkflowTaskPollers`: Defaults to 4

Author: TODO: this is actually overridden by SDK. server and SDK defaults differ for some reason.

Author: https://github.com/temporalio/temporal/blob/fe05751305b1cb50b68efa23f8aa5f1b34f45bc5/service/worker/service.go#L121

https://github.com/temporalio/sdk-java/blob/bef967639fcdbe14ca37b80ac816596412846e5f/temporal-sdk/src/main/java/io/temporal/worker/WorkerOptions.java#L51

```suggestion
Specifically, the primary scaling metrics are located in the server's dynamic configs:

- `MaxConcurrentActivityTaskPollers` and `MaxConcurrentWorkflowTaskPollers`: [Defaults to 4](https://github.com/temporalio/temporal/blob/fe05751305b1cb50b68efa23f8aa5f1b34f45bc5/service/worker/service.go#L121)
- `MaxConcurrentActivityExecutionSize` and ``MaxConcurrentWorkflowTaskExecutionSize`: [Defaults to 200](https://github.com/temporalio/sdk-java/blob/bef967639fcdbe14ca37b80ac816596412846e5f/temporal-sdk/src/main/java/io/temporal/worker/WorkerOptions.java#L51)
```
